### ISSUE TITLE
Vulnerability report

### ISSUE THREAD

**Issue by JafarAkhondali** [chcunningham/wc-talk#10]
We are a group of researchers from Leiden University, and we conduct research on vulnerabilities in open-source software. We have discovered and verified a high-severity vulnerability in your project(chcunningham/wc-talk). Explaining the vulnerability further in this issue could allow malicious users to access details, so we recommend [enabling private vulnerability reporting on GitHub](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/configuring-private-vulnerability-reporting-for-a-repository) to discuss this matter confidentially.
        After you have enabled this feature, please add a comment to this issue so we can continue our discussion. If you have any questions, feel free to leave a reply here or send an email to: j.akhoundali [at] liacs.leidenuniv.nl

**Comment by guest271314** [comment]
Wouldn't it make sense to publicly disclose the common vulnerabilities and exposures so we all can examine the issue?

**Comment by JafarAkhondali** [comment]
@guest271314 No, vulnerability reports should be discussed privately until maintainers provide a fix. Also if there is no response from the maintainers (in ~1month), then the vulnerability can be public so that users are aware of the issue.

**Comment by guest271314** [comment]
Interesting philosophy. I would just expose the exploit so users in the field can be aware that they are running software vulnerable to exploits and maintainers can also know they are producing code with vulnerabilities. Full disclosure all the way around. No secrets.

**Comment by JafarAkhondali** [comment]
We follow best practices from other source: https://cheatsheetseries.owasp.org/cheatsheets/Vulnerability_Disclosure_Cheat_Sheet.html
So If they don't respond in time, we will make the vulnerability details public(similar to Google's [Project Zero](https://googleprojectzero.blogspot.com/2020/01/policy-and-disclosure-2020-edition.html)).

**Comment by guest271314** [comment]
Right, the 2d option is full disclosure. It is possible that users and/or developers in the field can come up with a fix that maintainers have not thought about. Anyway, I'll check out what's going on here when the details are made public.

**Comment by JafarAkhondali** [comment]
@guest271314 Now you can check the details :)

**Comment by guest271314** [comment]
Reproducible with `node` v23.0.0-nightly202407272d1b4a8cf7 on Linux.

```
$ curl --path-as-is 0.0.0.0:8888/../../../../../../../../../../../../../../../../../../../etc/hostname
user
```

I converted the script to use Ecmascript Modules instead of CommonJS.

This also fixes the issue

```
  var filename = `./${new URL(request.url, import.meta.url).pathname}`;
```

by using the `base` parameter to `URL()` constructor https://developer.mozilla.org/en-US/docs/Web/API/URL/URL, `import.meta.url`, and prefixing the `filename` variable with `./`

```
Static file server running at
  => http://localhost:8888/index.html
CTRL + C to shutdown
.//etc/hostname
File doesn't exist:.//etc/hostname
```

```
$ curl --path-as-is 0.0.0.0:8888/../../../../../../../../../../../../../../../../../../../etc/hostname
404 Not Found
```
